### ISSUE TITLE
Eio_posix-based process runs into infinite loop after a few hours

### ISSUE THREAD

**Issue by smondet** [ocaml-multicore/eio#495]
I have long-running server using `Cohttp` and `Eio_posix` (because so far it uses `Eio_posix.Low_level.Process`). Once in a while after a few hours of running fine it apparently randomly turns into an infinite loop using 100% of one CPU.

`sudo strace -p <pid>` says (thousands of lines like this per second):

```default
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
[pid 3458264] rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
```

It's hard to reproduce reliably, let me know if there is anything I should to inspect the process better.

**Comment by smondet** [comment]
Yes before running into the problem `strace`'s output does look like that; it's only after a few hours that it turns into that infinite loop.

Not using multiple domains (although `Cohttp_eio` does seem to use the `#domain_mgr`), and not using any other library that deals with signals or I/O AFAIK.

**Comment by haesbaert** [comment]
https://github.com/ocaml/ocaml/blob/69472048a0d1782d2b773561588339136a8cf5b6/runtime/signals.c#L87

This is fishy indeed, this could happen if the bit in `caml_pending_signals` is set by the time we do a sigprocmask(SIG_BLOCK).

**Comment by smondet** [comment]
Thanks for the pointers, I'm trying 5.1.0-alpha1 now, I'll report when I have more info.

**Comment by smondet** [comment]
Haven't yet managed to build my project with 5.1𝛂 (need-to/will modify JSOO), but
in case it helps I put a signal handler to print the “current” stack-trace
while reproducing the infinite-loop, it looks like this:


```default
+signal-handler: -11
+backtrace:
+Raised by primitive operation at Eio_posix__Low_level.readv.(fun) in file "vendor/eio/lib_eio_posix/low_level.ml", line 140, characters 33-50
+Called from Eio_posix__Low_level.do_nonblocking in file "vendor/eio/lib_eio_posix/low_level.ml", line 31, characters 6-11
+
+signal-handler: -11
+backtrace:
+Raised by primitive operation at Eio__core__Cancel.Fiber_context.get_vars in file "vendor/eio/lib_eio/core/cancel.ml", line 214, characters 9-37
+Called from Eio__core__Fiber.get in file "vendor/eio/lib_eio/core/fiber.ml", line 280, characters 28-62
+Called from Eio__core__Debug.traceln in file "vendor/eio/lib_eio/core/debug.ml", line 29, characters 10-31
+
+signal-handler: -11
+backtrace:
+Raised by primitive operation at Eio__core__Cancel.Fiber_context.get_vars in file "vendor/eio/lib_eio/core/cancel.ml", line 214, characters 9-37
+Called from Eio__core__Fiber.get in file "vendor/eio/lib_eio/core/fiber.ml", line 280, characters 28-62
+Called from Eio__core__Debug.traceln in file "vendor/eio/lib_eio/core/debug.ml", line 29, characters 10-31
```

**Comment by talex5** [comment]
It would be interesting to attach gdb to it and step it a few times to find out why it's looping.

**Comment by talex5** [comment]
I've added a stress test for spawning processes in #519 - would be interesting to know if that triggers it (on either backend).

**Comment by talex5** [comment]
Thanks, this looks useful. From the initial stack trace, it seems this happened:

1. While doing some work (`Lf_queue.make`), OCaml noticed a signal was pending and called `caml_process_pending_signals_exn`.
2. That found the signal handler and used `caml_callback_exn` to call `camlEio_posix__Children__reap_nonblocking_677`.
3. That called `caml_unix_waitpid` (in non-blocking mode).
4. The C stub did `caml_enter_blocking_section`, which called `caml_process_pending_signals_exn` again.
5. That determined that a signal was pending and was just getting the signal mask.

Presumably, it's about to discover that the pending SIGCHLD signal is masked, since we're already in the handler. I would expect it to decide there's nothing to do and return, allowing the `waitpid` system call to run.

The steps show:
- It took the first signal word and found a signal was set.
- It started iterating through the bits.

Would be useful to see what happened after that. My guess is that it found the signal was masked, and returned to `caml_enter_blocking_section`, where it ran into this FIXME:

https://github.com/ocaml/ocaml/blob/e86f9e5d41217e8c824455206e854072b803b170/runtime/signals.c#L153-L155

In fact, it looks like this will just loop forever, which is what you're seeing.

(as a work-around, eio_linux now supports processes and doesn't use a signal handler)

**Comment by talex5** [comment]
I can reproduce this myself reliably now, and without Eio, by adding a small delay to OCaml's signal.h.
I've reported it at https://github.com/ocaml/ocaml/issues/12253.

**Comment by talex5** [comment]
This is fixed in OCaml 5.1.0-beta1.
